### ISSUE TITLE
PR 6 — Remaining model-type layers (Centroid, CenteredInstance, TopDown, BottomUp, Multi-class)

### ISSUE THREAD

**Issue by gitttt-1234** [talmolab/sleap-nn#514]
**Parent epic:** #508
**Prerequisites:** #509
**Blocks:** #516

---

## Goal

Ship the remaining five `InferenceLayer` subclasses following the pattern established by `SingleInstanceLayer` in #512:

1. `CentroidLayer` (includes `use_gt_centroids=True` mode)
2. `CenteredInstanceLayer`
3. `TopDownLayer` (composes Centroid + CenteredInstance)
4. `BottomUpLayer` (keeps the GPU/CPU split from current code)
5. `TopDownMultiClassLayer`
6. `BottomUpMultiClassLayer`

Each layer must pass a parity test against its current `InferenceModel` counterpart on its corresponding test fixture.

### `CentroidLayer`

```python
# sleap_nn/inference/layers/centroid.py

@attrs.define
class CentroidLayer(InferenceLayer):
    """Centroid detection: local peak finding, top-K by max_instances.

    Supports a `use_gt_centroids=True` mode for eval/ablation: skips the
    model forward, reads centroids from batch["instances"] via anchor_ind.
    """

    use_gt_centroids: bool = False
    anchor_ind: Optional[int] = None

    def postprocess(self, raw_out, info):
        if self.use_gt_centroids:
            # Short-circuit: no peak finding, use GT centroids from batch
            # See topdown.py:179-213 for current behavior
            ...
            return Outputs(pred_centroids=..., pred_centroid_values=...)

        if self.backend.does_baked_postproc:
            centroids = raw_out["peaks"]
            vals = raw_out["peak_vals"]
        else:
            confmaps = raw_out.get("CentroidConfmapsHead", next(iter(raw_out.values())))
            centroids, vals = find_local_peaks(
                confmaps,
                threshold=self.postprocess_config.peak_threshold,
                max_instances=self.postprocess_config.max_instances,
                refinement=self.postprocess_config.refinement,
            )

        centroids = self._reverse_coords(centroids, info)
        return Outputs(
            pred_centroids=centroids,
            pred_centroid_values=vals,
            preprocess_info=info,
        )
```

**Centroid-only note (issue #98):** `CentroidLayer` populates *only* `pred_centroids` / `pred_centroid_values` — not `pred_keypoints`. The Predictor decides how to package for `.slp` (see design doc §4B.4, "Centroid-only note").

### `CenteredInstanceLayer`

Preserves two current behaviors:
1. Model forward + global peak finding (default path)
2. **`use_gt_peaks=True`** — short-circuits model forward, matches predicted centroids to nearest GT instance, returns GT keypoints. Equivalent to today's `FindInstancePeaksGroundTruth`. Symmetric with `CentroidLayer(use_gt_centroids=True)`.

```python
# sleap_nn/inference/layers/centered_instance.py

@attrs.define
class CenteredInstanceLayer(InferenceLayer):
    """Top-down stage 2: keypoints within a centered crop.

    GT fallback: use_gt_peaks=True skips model forward, matches centroids
    against GT instances (nearest-neighbor by keypoint distance), returns
    GT keypoints. Preserves FindInstancePeaksGroundTruth semantics.
    """

    use_gt_peaks: bool = False
    """When True, bypass model forward and return GT keypoints for each centroid.
    Requires the batch to carry GT 'instances'. Raises UnsupportedForBackend
    if backend is ONNX/TRT — GT fallback has no model pass to export."""

    def predict(self, crops, centroids=None, gt_instances=None):
        if self.use_gt_peaks:
            if self.backend.does_baked_postproc:
                raise UnsupportedForBackend(
                    "use_gt_peaks=True is Torch-only; GT matching has no ONNX path"
                )
            return self._gt_peaks(centroids, gt_instances)
        return super().predict(crops)

    def _gt_peaks(self, centroids, gt_instances) -> Outputs:
        """Match centroids to nearest GT instance, return GT keypoints.

        Preserves the logic in current topdown.py:339-470
        (FindInstancePeaksGroundTruth.forward).
        """
        # Nearest-neighbor match centroids → GT instances (min sum-squared-dist
        # over nodes), return matched GT keypoints with vals=1.0 for valid.
        ...

    def postprocess(self, raw_out, info):
        confmaps = raw_out.get("CenteredInstanceConfmapsHead", ...)
        peaks, peak_vals = find_global_peaks(confmaps, ...)
        # Peaks stay in CROP-local coordinates. TopDownLayer un-crops.
        peaks = undo_stride(peaks, info.output_stride)
        peaks = undo_input_scale(peaks, info.input_scale)
        return Outputs(
            pred_keypoints=peaks.unsqueeze(1),
            pred_peak_values=peak_vals.unsqueeze(1),
            instance_scores=peak_vals.mean(dim=-1, keepdim=True),
            preprocess_info=info,
        )
```

### `TopDownLayer` (composition, with three-stage filter orchestration)

Top-down composes two layers AND owns the between-stage filter logic. Per design doc §4.6, there are three filter stages:

| Stage | Owner | Filters |
|---|---|---|
| **A** — inside stage 1 | `CentroidLayer.postprocess_config` | `peak_threshold`, `max_instances` |
| **B** — between stages | `TopDownLayer.predict()` (this PR) | auto NaN-prune, opt-in centroid NMS |
| **C** — final outputs | `Predictor` via `FilterPipeline` | full `FilterConfig` |

Stage B is **new work** in this PR. Today's code has no pruning between stages — stage 2 runs on every `max_instances` slot including NaN-padded ones, wasting compute. #514 adds:

1. **Automatic NaN-centroid pruning** (Torch path only; ONNX bakes fixed shapes and cannot prune mid-graph)
2. **Optional `centroid_nms`** for close-together animals — dedupe overlapping centroids by bbox-IoU *before* stage 2, not after

```python
# sleap_nn/inference/layers/topdown.py

@attrs.define
class TopDownLayer(InferenceLayer):
    """Two-stage top-down with explicit Stage-B filter orchestration."""

    centroid_layer: CentroidLayer = attrs.field()
    instance_layer: CenteredInstanceLayer = attrs.field()
    use_gt_centroids: bool = False

    # Stage B knobs (new in this PR)
    centroid_nms: bool = False
    centroid_nms_threshold: float = 0.5

    def predict(self, x):
        # ── Stage 1 ──
        if self.use_gt_centroids:
            centroid_outputs = self._gt_centroids(x)
        else:
            centroid_outputs = self.centroid_layer.predict(x)
        if centroid_outputs.n_instances == 0:
            return Outputs()

        # ── Stage B.1: auto NaN-prune (Torch only) ──
        if not self.centroid_layer.backend.does_baked_postproc:
            centroid_outputs = _drop_nan_centroids(centroid_outputs)

        # ── Stage B.2: optional centroid NMS (opt-in) ──
        if self.centroid_nms and not self.centroid_layer.backend.does_baked_postproc:
            centroid_outputs = _centroid_nms(
                centroid_outputs,
                threshold=self.centroid_nms_threshold,
                crop_hw=self.instance_layer.preprocess_config.crop_size,
            )
        if centroid_outputs.n_instances == 0:
            return Outputs()

        # ── Crop extraction + stage 2 (only survivors) ──
        processed, _ = self.centroid_layer.preprocess(x)
        crops, bboxes = extract_instance_crops(
            processed, centroid_outputs.pred_centroids,
            crop_hw=self.instance_layer.preprocess_config.crop_size,
        )
        instance_outputs = self.instance_layer.predict(crops)

        # ── Un-crop ──
        crop_topleft = bboxes[:, 0, :]
        instance_outputs = attrs.evolve(
            instance_outputs,
            pred_keypoints=add_crop_offset(instance_outputs.pred_keypoints, crop_topleft),
            pred_centroids=centroid_outputs.pred_centroids,
            pred_centroid_values=centroid_outputs.pred_centroid_values,
            instance_bboxes=bboxes,
        )
        return instance_outputs

    def postprocess(self, raw_out, info):
        raise NotImplementedError("TopDownLayer uses predict(), not postprocess()")

    def warmup(self, sample_shape=None):
        self.centroid_layer.warmup(sample_shape)
        crop_hw = self.instance_layer.preprocess_config.crop_size or (96, 96)
        self.instance_layer.warmup((1, self.input_channels, *crop_hw))


# New helpers in this PR — belong in ops/filters.py or inline in topdown.py
def _drop_nan_centroids(centroid_outputs: Outputs) -> Outputs:
    """Drop slots where pred_centroids is all-NaN.

    Torch-path optimization. CentroidLayer NaN-pads below-threshold and
    over-cap slots (to satisfy fixed-size output convention). This prunes
    them before crop extraction so stage 2 doesn't run on garbage crops.
    """
    centroids = centroid_outputs.pred_centroids      # (B, I, 2)
    valid_mask = ~centroids.isnan().any(dim=-1)      # (B, I)
    # For now, simplest impl: reshape (B, I, ...) → (valid_count, ...)
    # Assumes uniform validity per batch for simplicity; handle padded
    # batches by keeping max valid count and re-NaN-padding the shorter ones.
    ...

def _centroid_nms(
    centroid_outputs: Outputs, threshold: float, crop_hw: Tuple[int, int],
) -> Outputs:
    """Greedy NMS on centroids by bbox IoU.

    For each centroid, compute the bbox that WOULD be used for cropping.
    Sort by centroid_value descending; greedy-drop anything with IoU >= threshold
    against an already-kept bbox.
    """
    ...
```

### `BottomUpLayer`

Preserves the GPU/CPU split from the current code (`bottomup.py:126-286`):

```python
# sleap_nn/inference/layers/bottomup.py

@attrs.define
class BottomUpLayer(InferenceLayer):
    """Multi-instance with PAF-based grouping.

    Keeps the internal GPU→CPU split for pipelinability:
      - GPU: model forward, local peaks, PAF line scoring
      - CPU: Hungarian + BFS grouping (PAFScorer.predict)
    """

    paf_scorer: PAFScorer = attrs.field()

    def postprocess(self, raw_out, info):
        # GPU stage
        cms = raw_out.get("MultiInstanceConfmapsHead")
        pafs = raw_out.get("PartAffinityFieldsHead").permute(0, 2, 3, 1)
        peaks, peak_vals, peak_channel_inds = find_local_peaks(cms.detach(), ...)

        # Early-exit for noisy under-trained models (per §3.3)
        if self.postprocess_config.paf.max_peaks_per_node:
            if peaks.shape[1] > self.postprocess_config.paf.max_peaks_per_node:
                return Outputs()  # empty prediction with warning

        # CPU stage (or worker pool via #517)
        predicted_instances = self.paf_scorer.predict(
            pafs, peaks, peak_vals, peak_channel_inds,
        )
        predicted_instances = undo_input_scale(predicted_instances, info.input_scale)
        predicted_instances = undo_eff_scale(predicted_instances, info.eff_scale)

        return Outputs(
            pred_keypoints=predicted_instances,
            pred_peak_values=...,
            pred_confmaps=cms if self.postprocess_config.return_confmaps else None,
            pred_pafs=pafs if self.postprocess_config.return_pafs else None,
            pred_paf_graph=(peaks, edge_inds, ...) if self.postprocess_config.return_paf_graph else None,
            preprocess_info=info,
        )
```

### `TopDownMultiClassLayer`

Extends `TopDownLayer` with per-node class decoding:

```python
# sleap_nn/inference/layers/topdown_mc.py

@attrs.define
class TopDownMultiClassLayer(TopDownLayer):
    """Top-down with per-instance identity classification.

    Adds per-node class index + optional full logits to Outputs.
    """
    # Reuses TopDownLayer.predict(); modifies stage-2 postprocess to extract
    # class vectors from the instance_model's output dict.
    ...
```

### `BottomUpMultiClassLayer`

```python
# sleap_nn/inference/layers/bottomup_mc.py

@attrs.define
class BottomUpMultiClassLayer(BottomUpLayer):
    """Bottom-up with class-map-based identity (no PAF grouping).

    Replaces PAFScorer.predict() with classify_peaks_from_maps().
    """
    ...
```

## Parity testing — critical gate

Every layer must produce byte-for-byte identical outputs vs the current code on its fixture checkpoint:

```python
# tests/inference/layers/test_all_layers_parity.py

@pytest.mark.parametrize("model_type,layer_cls,fixture_path", [
    ("centroid", CentroidLayer, "centroid/"),
    ("centered_instance", CenteredInstanceLayer, "centered_instance/"),
    ("topdown", TopDownLayer, ["centroid/", "centered_instance/"]),
    ("bottomup", BottomUpLayer, "bottomup/"),
    ("topdown_mc", TopDownMultiClassLayer, ["centroid/", "topdown_mc/"]),
    ("bottomup_mc", BottomUpMultiClassLayer, "bottomup_mc/"),
])
def test_layer_parity(model_type, layer_cls, fixture_path):
    # Load via new path
    layer = layer_cls.from_checkpoint(fixture_path, device="cpu")
    # Load via old path
    old_model = load_old_inference_model(model_type, fixture_path)
    # Compare on shared input
    image = load_test_image()
    new_output = layer.predict(image)
    old_output = old_model({"image": image})
    _assert_output_parity(new_output, old_output, tol=1e-5)
```

### All layers
- [ ] All 6 layer classes ship with full `predict()` / `postprocess()` implementations
- [ ] `use_gt_centroids=True` mode works on `CentroidLayer` (covers `FindInstancePeaksGroundTruth` use cases for stage-1-oracle eval)
- [ ] Parity tests pass for every model type (byte-for-byte vs current code, tol=1e-5)
- [ ] `BottomUpLayer.postprocess` preserves the GPU/CPU split documented in §4.7
- [ ] All 5 `return_*` flags populate the correct `Outputs` fields (per §3.8 field table in §4B.4)
- [ ] `max_peaks_per_node` early-exit works for bottom-up (current behavior preserved)
- [ ] Each layer's `.warmup()` works on CPU/MPS

### TopDownLayer Stage-B filtering (new in this PR)
- [ ] **NaN-centroid auto-prune**: verify via a mock `instance_layer` that counts forward calls — stage 2 is invoked with only the valid centroid count, not `max_instances` padded slots. Tested on a fixture with half-NaN centroids.
- [ ] **ONNX path unchanged**: when `centroid_layer.backend.does_baked_postproc=True`, Stage B is skipped in Python (the graph handled it). End-to-end outputs match Torch path within tolerance.
- [ ] **Opt-in centroid NMS**: on a close-overlap fixture (two centroids on one animal), `centroid_nms=True` drops one centroid before stage 2 — verified via call count on mock `instance_layer`. With `centroid_nms=False` (default), both centroids pass through.
- [ ] **Parity with filters off**: `TopDownLayer(centroid_nms=False)` matches current top-down output byte-for-byte on all fixture checkpoints. The auto NaN-prune is a pure optimization and must not change final results.
- [ ] **Filter stage separation documented**: a unit test asserts the three-stage ownership (Stage A filters change `CentroidLayer.postprocess_config.peak_threshold`; Stage B filters change `TopDownLayer.centroid_nms`; Stage C filters change `Predictor.filter_config`) produce the expected disjoint behaviors.

### Shared
- [ ] `ruff check` + `black --check` clean

## Files touched

**Created:**
- `sleap_nn/inference/layers/centroid.py`
- `sleap_nn/inference/layers/centered_instance.py`
- `sleap_nn/inference/layers/topdown.py`
- `sleap_nn/inference/layers/bottomup.py`
- `sleap_nn/inference/layers/topdown_mc.py`
- `sleap_nn/inference/layers/bottomup_mc.py`
- `tests/inference/layers/test_all_layers_parity.py`
- Per-layer focused tests (e.g. `test_centroid_layer.py`, `test_bottomup_layer.py`)

**Modified:**
- `sleap_nn/inference/layers/backends/loader.py` — extend `load_layer` dispatch to all 6 types

## Locked decisions for this PR

**`FindInstancePeaksGroundTruth` preserved as `CenteredInstanceLayer(use_gt_peaks=True)`.** Design doc §7 item 6 (revised). Rationale: the refactor's core principle is parity-first — no user-facing behavior change in PRs 1-13. Migrating the surface to an evaluation module during the refactor muddies parity testing. Evaluation-module migration deferred to a post-refactor PR.

Acceptance criteria specific to GT-fallback preservation:

- [ ] `CenteredInstanceLayer(use_gt_peaks=True).predict(crops, centroids, gt_instances)` produces byte-for-byte identical outputs to today's `FindInstancePeaksGroundTruth.forward()` — verified on the fixture that exercises GT fallback
- [ ] `TopDownLayer` composition works with `CentroidLayer(use_gt_centroids=True)` + `CenteredInstanceLayer(use_gt_peaks=True)` — the full GT-oracle path
- [ ] Attempting `CenteredInstanceLayer(use_gt_peaks=True)` with an ONNX/TRT backend raises `UnsupportedForBackend` with a clear message ("GT fallback is Torch-only; no ONNX path")
- [ ] Existing callers of `FindInstancePeaksGroundTruth` (scripts, tests, eval code) have a documented migration path — either via compat shim in #520 or via direct import from the new layer location

## Out of scope

- ONNX/TensorRT backends (#515)
- PAF worker pool for bottom-up (#517)
- Legacy SLEAP ≤1.4 checkpoint loading (#520)
- **Centroid-only `.slp` output (issue #98)** — deferred to **#522**. `CentroidLayer.predict()` in this PR returns structured `Outputs` with `pred_centroids` only; #522 handles how those get packaged into a saveable `.slp` file.
- Moving `FindInstancePeaksGroundTruth` to `evaluation/oracle.py` — deferred to a post-refactor PR if desired.

**Comment by gitttt-1234** [comment]
Closed — merged as part of the inference refactor stack (epic #508).
